### PR TITLE
Add new target API to get a target relative to a specific BaseModule.

### PR DESCRIPTION
**Pull request by mikeurbach**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Feature (or new API)

The first commit contains some internal refactoring of how ids get named in a RawModule, which was required for the new API. The ids generated normally are named, then any `atModuleBodyEnd` blocks are evaluated and the RawModule is closed, and finally any ids generated by `atModuleBodyEnd` blocks are named. This allows the new `toRelativeTarget` to be called within an `atModuleBodyEnd` block for the root RawModule.

The second commit actually adds the new APIs.

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference).

#### Release Notes

The new toRelativeTarget API augments the existing toAbsoluteTarget API, to provide a mechanism to get hierarchical paths to a NamedComponent relative to a known root.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.5.x`, `3.6.x`, or `5.x` depending on impact, API modification or big change: `6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/test/scala/chisel3/internal/NameCollisionSpec.scala
Comment (mikeurbach) [inline src/test/scala/chisel3/internal/NameCollisionSpec.scala:75]: This ended up being the only change I could see related to the two-phase HasId naming in RawModule. Now that we are doing a pass of naming before `checkPorts`, the error message from check ports doesn't come up in this case. We do still get a nice error that basically says the same thing, so I think this is fine.

File: core/src/main/scala/chisel3/RawModule.scala
Comment (jackkoenig) [inline core/src/main/scala/chisel3/RawModule.scala:178]:
```suggestion
    for (id <- _ids.view.drop(numInitialIds)) {
      nameId(id)
```
This probably won't matter much in practice, but that `.drop` does a full copy of the `ArrayBuffer`, whereas first asking for a `.view` then makes the `drop` lazy such that it won't materialize a new datastructure and instead will just let us iterate.

I experimented with a few alternatives likes so:
```scala
//> using scala "2.13.11"
//> using options "-unchecked", "-deprecation", "-language:reflectiveCalls", "-feature", "-Xcheckinit", "-Xfatal-warnings", "-Ywarn-dead-code", "-Ywarn-unused", "-Ymacro-annotations"

import scala.collection.mutable

object logEvent {
  private val totals = new mutable.HashMap[String, Double]()
  def apply[R](name: String)(block: => R): R = {
    val t0 = System.nanoTime()
    val result = block
    val t1 = System.nanoTime()
    val timeMillis = (t1 - t0) / 1000000.0

    val total = timeMillis + totals.getOrElse(name, 0.0)
    totals(name) = total
    println(f"$name took $timeMillis%.1f ms (total $total%.1f ms)")
    result
  }
}

object Main extends App {
  val array = logEvent("Make ArrayBuffer") {
    val xs = mutable.ArrayBuffer.empty[String]
    for (i <- 0 until 100000000) {
      xs += s"value_$i"
    }
    xs
  }

  for (_ <- 0 until 10) {

    logEvent(".drop") {
      var i = 0L
      for (x <- array.drop(array.size - 200000)) {
        i += x.size
      }
      println(i)
    }

    logEvent(".view.drop") {
      var i = 0L
      for (x <- array.view.drop(array.size - 200000)) {
        i += x.size
      }
      println(i)
    }

    logEvent(".iterator.drop") {
      var i = 0L
      for (x <- array.iterator.drop(array.size - 200000)) {
        i += x.size
      }
      println(i)
    }

    logEvent("using indexing") {
      var i = 0L
      var idx = array.size - 200000
      while (idx < array.size) {
        i += array(idx).size
        idx += 1
      }
      println(i)
    }
  }
}
```

File: src/test/scala/chiselTests/ToTargetSpec.scala
Comment (jackkoenig) [inline src/test/scala/chiselTests/ToTargetSpec.scala:111]: Both of these tests have the top module as the one you're relative to (even though one uses `Some` vs. `None` as the argument). It would be good if one of the relative tests were relative to a module that is not the top.

**Review by jackkoenig — APPROVED** [review]
Generally looks good but please see the suggestions before merging.

File: src/test/scala/chiselTests/ToTargetSpec.scala
Comment (mikeurbach) [reply src/test/scala/chiselTests/ToTargetSpec.scala:111]: 🤦 yeah, thanks.

File: core/src/main/scala/chisel3/RawModule.scala
Comment (mikeurbach) [reply core/src/main/scala/chisel3/RawModule.scala:178]: ah, thanks for the experimentation and pointers! btw, is there any canonical way to know this without reading the source or doing experiments? i didn't see any mention of these concerns here: https://www.scala-lang.org/api/2.13.3/scala/collection/mutable/ArrayBuffer.html#drop(n:Int):C

File: core/src/main/scala/chisel3/internal/Builder.scala
Comment (mikeurbach) [reply core/src/main/scala/chisel3/internal/Builder.scala:400]: Sounds good... I was banging my head trying to get this working and cut out the view logic. I'll add a test and handle views similarly to toAbsoluteTarget.

File: core/src/main/scala/chisel3/Module.scala
Comment (jackkoenig) [inline core/src/main/scala/chisel3/Module.scala:631]: I feel like we could simplify this logic a bit, several of the cases seem basically duplicated, especially lines 619 and 626 as well as 621 and 628.
```scala
if (root.contains(this)) getTarget
else if (this == ViewParent) ViewParent.absoluteTarget
else (root, this._parent) match {
  case (_, Some(parent)) =>  parent.toRelativeTarget(root).instOf(this.instanceName, name)
  case (Some(definedRoot), None) =>
    throwException(s"Requested .toRelativeTarget relative to ${definedRoot.name}, but it is not an ancestor")
  case (None, None) => getTarget
}
```
Maybe? I just sketched that up really fast so it may not be correct

Comment (jackkoenig) [reply core/src/main/scala/chisel3/Module.scala:631]: I'll leave it to you to decide how to express this in a way that is the most clear, approved for whatever you prefer.

**Review by jackkoenig — APPROVED** [review]

File: core/src/main/scala/chisel3/Module.scala
Comment (mikeurbach) [reply core/src/main/scala/chisel3/Module.scala]: Your suggestion works great as-is, just applied my old comments to the new structure. Thanks for taking the time to think hard enough about this to make that suggestion.

**Pull request merged**
